File: speech-to-text/websocket/streaming-text-output-transcription/python/transcribe.py
#!/usr/bin/env python3
"""
Smallest AI Speech-to-Text - Streaming Transcription

Stream an audio file through the WebSocket API and get transcription responses.
All transcripts are appended to a text file. Console shows only final transcripts.

Usage: python transcribe.py <audio_file>

Output:
- Console shows final transcripts (is_final=true)
- {filename}_responses.txt - All transcripts as plain text
"""

import asyncio
import json
import os
import sys
from pathlib import Path
from urllib.parse import urlencode

import librosa
import numpy as np
import websockets

WS_URL = "wss://waves-api.smallest.ai/api/v1/lightning/get_text"
OUTPUT_DIR = "."

# The following are all the features supported by the WebSocket endpoint (Streaming API)
LANGUAGE = "hi"
ENCODING = "linear16"
SAMPLE_RATE = 16000
WORD_TIMESTAMPS = False
FULL_TRANSCRIPT = True
SENTENCE_TIMESTAMPS = False
DIARIZE = False
REDACT_PII = False
REDACT_PCI = False
NUMERALS = "auto"
KEYWORDS = []


async def transcribe(audio_file: str, api_key: str, on_response):
    params = {
        "language": LANGUAGE,
        "encoding": ENCODING,
        "sample_rate": SAMPLE_RATE,
        "word_timestamps": str(WORD_TIMESTAMPS).lower(),
        "full_transcript": str(FULL_TRANSCRIPT).lower(),
        "sentence_timestamps": str(SENTENCE_TIMESTAMPS).lower(),
        "diarize": str(DIARIZE).lower(),
        "redact_pii": str(REDACT_PII).lower(),
        "redact_pci": str(REDACT_PCI).lower(),
        "numerals": str(NUMERALS).lower() if isinstance(NUMERALS, bool) else NUMERALS,
    }
    if KEYWORDS:
        params["keywords"] = json.dumps(KEYWORDS)

    url = f"{WS_URL}?{urlencode(params)}"
    headers = {"Authorization": f"Bearer {api_key}"}

    audio, _ = librosa.load(audio_file, sr=SAMPLE_RATE, mono=True)
    chunk_duration = 0.1
    chunk_size = int(chunk_duration * SAMPLE_RATE)

    async with websockets.connect(url, additional_headers=headers) as ws:
        async def send_audio():
            for i in range(0, len(audio), chunk_size):
                chunk = audio[i:i + chunk_size]
                pcm16 = (chunk * 32768.0).astype(np.int16).tobytes()
                await ws.send(pcm16)
                await asyncio.sleep(chunk_duration)
            await ws.send(json.dumps({"type": "end"}))

        async def receive_responses():
            async for message in ws:
                result = json.loads(message)
                on_response(result)
                if result.get("is_last"):
                    break

        await asyncio.gather(send_audio(), receive_responses())


# This function is designed to process feature outputs for all the features supported
# by the WebSocket endpoint (Streaming API)
def process_response(result: dict, output_file: Path):
    transcript = result.get("transcript", "")
    if transcript:
        with open(output_file, "a", encoding="utf-8") as f:
            f.write(transcript + "\n")

    if result.get("is_final"):
        transcript = result.get("transcript", "")
        if result.get("is_last"):
            print(f"[FINAL] {transcript}")

            if result.get("full_transcript"):
                print("\n" + "=" * 60)
                print("FULL TRANSCRIPT")
                print("=" * 60)
                print(result.get("full_transcript"))

            # Language
            if result.get("language"):
                print("\n" + "-" * 60)
                print("LANGUAGE")
                print("-" * 60)
                print(f"Detected: {result.get('language')}")
                if result.get("languages"):
                    print(f"All: {result.get('languages')}")

            # Utterances (speaker diarization)
            if result.get("utterances"):
                print("\n" + "-" * 60)
                print("UTTERANCES")
                print("-" * 60)
                for utt in result["utterances"]:
                    speaker = utt.get("speaker", "")
                    start = utt.get("start", 0)
                    end = utt.get("end", 0)
                    text = utt.get("text", "")
                    if speaker != "":
                        print(f"[{start:.2f}s - {end:.2f}s] speaker_{speaker}: {text}")
                    else:
                        print(f"[{start:.2f}s - {end:.2f}s] {text}")

            # Word timestamps
            if result.get("words"):
                print("\n" + "-" * 60)
                print("WORD TIMESTAMPS")
                print("-" * 60)
                for word in result["words"]:
                    start = word.get("start", 0)
                    end = word.get("end", 0)
                    text = word.get("word", "")
                    confidence = word.get("confidence", 0)
                    speaker = word.get("speaker", "")
                    if speaker != "":
                        print(f"[{start:.2f}s - {end:.2f}s] speaker_{speaker}: {text} ({confidence:.2f})")
                    else:
                        print(f"[{start:.2f}s - {end:.2f}s] {text} ({confidence:.2f})")

            # Redacted entities
            if result.get("redacted_entities"):
                print("\n" + "-" * 60)
                print("REDACTED ENTITIES")
                print("-" * 60)
                for entity in result["redacted_entities"]:
                    print(f"  {entity}")

            print("\n" + "=" * 60)
        else:
            print(f"{transcript}", end="", flush=True)


def main():
    if len(sys.argv) < 2:
        print("Usage: python transcribe.py <audio_file>")
        sys.exit(1)

    audio_file = sys.argv[1]
    api_key = os.environ.get("SMALLEST_API_KEY")

    if not api_key:
        print("Error: SMALLEST_API_KEY environment variable not set")
        sys.exit(1)

    audio_path = Path(audio_file)
    if not audio_path.exists():
        print(f"Error: File not found: {audio_file}")
        sys.exit(1)

    output_dir = Path(OUTPUT_DIR)
    output_dir.mkdir(parents=True, exist_ok=True)
    output_file = output_dir / f"{audio_path.stem}_responses.txt"

    if output_file.exists():
        output_file.unlink()

    print(f"Streaming: {audio_path.name}")
    print(f"Responses saved to: {output_file}")
    print("-" * 60)

    asyncio.run(transcribe(
        audio_file,
        api_key,
        lambda result: process_response(result, output_file)
    ))

    print("-" * 60)
    print(f"Done!")


if __name__ == "__main__":
    main()

File: speech-to-text/websocket/streaming-text-output-transcription/javascript/transcribe.js
#!/usr/bin/env node
/**
 * Smallest AI Speech-to-Text - Streaming Transcription
 *
 * Stream an audio file through the WebSocket API and get transcription responses.
 * All transcripts are appended to a text file. Console shows only final transcripts.
 *
 * Usage: node transcribe.js <audio_file>
 *
 * Output:
 * - Console shows final transcripts (is_final=true)
 * - {filename}_responses.txt - All transcripts as plain text
 */

const fs = require("fs");
const path = require("path");
const WebSocket = require("ws");
const wav = require("wav");

const WS_URL = "wss://waves-api.smallest.ai/api/v1/lightning/get_text";
const OUTPUT_DIR = ".";

// The following are all the features supported by the WebSocket endpoint (Streaming API)
const LANGUAGE = "en";
const ENCODING = "linear16";
const SAMPLE_RATE = 16000;
const WORD_TIMESTAMPS = false;
const FULL_TRANSCRIPT = true;
const SENTENCE_TIMESTAMPS = false;
const DIARIZE = false;
const REDACT_PII = false;
const REDACT_PCI = false;
const NUMERALS = "auto";
const KEYWORDS = [];


async function loadAudio(audioFile) {
  return new Promise((resolve, reject) => {
    const reader = new wav.Reader();
    const chunks = [];

    reader.on("format", (format) => {
      reader.on("data", (chunk) => chunks.push(chunk));
      reader.on("end", () => {
        const buffer = Buffer.concat(chunks);
        const samples = new Int16Array(buffer.buffer, buffer.byteOffset, buffer.length / 2);

        // Resample if needed
        if (format.sampleRate !== SAMPLE_RATE) {
          const ratio = format.sampleRate / SAMPLE_RATE;
          const newLength = Math.floor(samples.length / ratio);
          const resampled = new Int16Array(newLength);
          for (let i = 0; i < newLength; i++) {
            resampled[i] = samples[Math.floor(i * ratio)];
          }
          resolve(resampled);
        } else {
          resolve(samples);
        }
      });
    });

    reader.on("error", reject);
    fs.createReadStream(audioFile).pipe(reader);
  });
}


async function transcribe(audioFile, apiKey, onResponse) {
  const params = new URLSearchParams({
    language: LANGUAGE,
    encoding: ENCODING,
    sample_rate: SAMPLE_RATE,
    word_timestamps: WORD_TIMESTAMPS,
    full_transcript: FULL_TRANSCRIPT,
    sentence_timestamps: SENTENCE_TIMESTAMPS,
    diarize: DIARIZE,
    redact_pii: REDACT_PII,
    redact_pci: REDACT_PCI,
    numerals: NUMERALS,
  });

  if (KEYWORDS.length > 0) {
    params.append("keywords", JSON.stringify(KEYWORDS));
  }

  const url = `${WS_URL}?${params}`;
  const audio = await loadAudio(audioFile);
  const chunkDuration = 0.1;
  const chunkSize = Math.floor(chunkDuration * SAMPLE_RATE);

  return new Promise((resolve, reject) => {
    const ws = new WebSocket(url, {
      headers: { Authorization: `Bearer ${apiKey}` },
    });

    ws.on("open", async () => {
      for (let i = 0; i < audio.length; i += chunkSize) {
        const chunk = audio.slice(i, i + chunkSize);
        ws.send(Buffer.from(chunk.buffer, chunk.byteOffset, chunk.byteLength));
        await new Promise((r) => setTimeout(r, chunkDuration * 1000));
      }
      ws.send(JSON.stringify({ type: "end" }));
    });

    ws.on("message", (data) => {
      const result = JSON.parse(data.toString());
      onResponse(result);
      if (result.is_last) {
        ws.close();
      }
    });

    ws.on("close", resolve);
    ws.on("error", reject);
  });
}


// This function is designed to process feature outputs for all the features supported
// by the WebSocket endpoint (Streaming API)
function processResponse(result, outputFile) {
  const transcript = result.transcript || "";
  if (transcript) {
    fs.appendFileSync(outputFile, transcript + "\n", "utf-8");
  }

  if (result.is_final) {
    if (result.is_last) {
      console.log(`[FINAL] ${transcript}`);

      if (result.full_transcript) {
        console.log("\n" + "=".repeat(60));
        console.log("FULL TRANSCRIPT");
        console.log("=".repeat(60));
        console.log(result.full_transcript);
      }

      // Language
      if (result.language) {
        console.log("\n" + "-".repeat(60));
        console.log("LANGUAGE");
        console.log("-".repeat(60));
        console.log(`Detected: ${result.language}`);
        if (result.languages) {
          console.log(`All: ${JSON.stringify(result.languages)}`);
        }
      }

      // Utterances (speaker diarization)
      if (result.utterances && result.utterances.length > 0) {
        console.log("\n" + "-".repeat(60));
        console.log("UTTERANCES");
        console.log("-".repeat(60));
        for (const utt of result.utterances) {
          const speaker = utt.speaker !== undefined ? utt.speaker : "";
          const start = utt.start || 0;
          const end = utt.end || 0;
          const text = utt.text || "";
          if (speaker !== "") {
            console.log(`[${start.toFixed(2)}s - ${end.toFixed(2)}s] speaker_${speaker}: ${text}`);
          } else {
            console.log(`[${start.toFixed(2)}s - ${end.toFixed(2)}s] ${text}`);
          }
        }
      }

      // Word timestamps
      if (result.words && result.words.length > 0) {
        console.log("\n" + "-".repeat(60));
        console.log("WORD TIMESTAMPS");
        console.log("-".repeat(60));
        for (const word of result.words) {
          const start = word.start || 0;
          const end = word.end || 0;
          const text = word.word || "";
          const confidence = word.confidence || 0;
          const speaker = word.speaker !== undefined ? word.speaker : "";
          if (speaker !== "") {
            console.log(`[${start.toFixed(2)}s - ${end.toFixed(2)}s] speaker_${speaker}: ${text} (${confidence.toFixed(2)})`);
          } else {
            console.log(`[${start.toFixed(2)}s - ${end.toFixed(2)}s] ${text} (${confidence.toFixed(2)})`);
          }
        }
      }

      // Redacted entities
      if (result.redacted_entities && result.redacted_entities.length > 0) {
        console.log("\n" + "-".repeat(60));
        console.log("REDACTED ENTITIES");
        console.log("-".repeat(60));
        for (const entity of result.redacted_entities) {
          console.log(`  ${entity}`);
        }
      }

      console.log("\n" + "=".repeat(60));
    } else {
      process.stdout.write(transcript);
    }
  }
}


async function main() {
  if (process.argv.length < 3) {
    console.log("Usage: node transcribe.js <audio_file>");
    process.exit(1);
  }

  const audioFile = process.argv[2];
  const apiKey = process.env.SMALLEST_API_KEY;

  if (!apiKey) {
    console.error("Error: SMALLEST_API_KEY environment variable not set");
    process.exit(1);
  }

  if (!fs.existsSync(audioFile)) {
    console.error(`Error: File not found: ${audioFile}`);
    process.exit(1);
  }

  const audioPath = path.parse(audioFile);
  const outputFile = path.join(OUTPUT_DIR, `${audioPath.name}_responses.txt`);

  if (fs.existsSync(outputFile)) {
    fs.unlinkSync(outputFile);
  }

  console.log(`Streaming: ${audioPath.base}`);
  console.log(`Responses saved to: ${outputFile}`);
  console.log("-".repeat(60));

  try {
    await transcribe(audioFile, apiKey, (result) => processResponse(result, outputFile));
    console.log("-".repeat(60));
    console.log("Done!");
  } catch (error) {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  }
}

main();
